### PR TITLE
Adds solo_gem and extra_gem rake tasks for appliance_installation_script

### PR DESCRIPTION
**Pull request by NickLaMuro**
Description
-----------
Allows the installation of gems that are not `stackprof`/`manageiq-performance`, alongside with or excluding those gems.

Built off of https://github.com/ManageIQ/manageiq-performance/pull/24

TODO
----
* [x] Include a `extra_c_gem` or `extra_c_ext_gem` task
* [x] Include a `solo_c_gem` or `solo_c_ext_gem` task"

~~* [ ] Have the `include_stackprof` task call out to a generic `extra_c_gem` task~~

Will do this last one in another pass I think.  Would rather have tests in place before doing that.

Links
-----
* Builds off of https://github.com/ManageIQ/manageiq-performance/pull/24

### CODE REVIEW

**Comment by kbrock** [comment]
Huh, thought you already merged this a while ago.

**Comment by NickLaMuro** [comment]
@kbrock Can't, based off of https://github.com/ManageIQ/manageiq-performance/pull/24

File: tasks/support/templates/appliance_installation_script.rb.erb [hunk 1]
```diff
@@ -7,22 +7,34 @@ require "bundler/definition"
 
 GEM_INSTALL_DIR          = "/tmp/manageiq-performance-gem-dir"
 
+<% unless @solo_gem -%>
```
Comment (kbrock) [inline tasks/support/templates/appliance_installation_script.rb.erb]: would it logically make more sense to put this `unless` check around this and the next block?

File: tasks/support/templates/appliance_installation_script.rb.erb [hunk 2]
```diff
@@ -7,22 +7,34 @@ require "bundler/definition"
...
 
-<% if @stackprof_gemspec and @stackprof_gem_base64_string -%>
+<% if @stackprof_gemspec and @stackprof_gem_base64_string and not @solo_gem -%>
 STACKPROF_GEM_NAME         = "<%= @stackprof_gemspec.name %>"
 STACKPROF_GEM_VERSION      = "<%= @stackprof_gemspec.version %>"
 STACKPROF_FULL_GEM_NAME    = "#{STACKPROF_GEM_NAME}-#{STACKPROF_GEM_VERSION}"
 STACKPROF_INSTALL_LOCATION = "#{GEM_INSTALL_DIR}/gems/#{STACKPROF_FULL_GEM_NAME}"
 <% end -%>
 
+<% if @other_gems -%>
```
Comment (kbrock) [inline tasks/support/templates/appliance_installation_script.rb.erb]: empty array here if there are no others?

File: tasks/support/templates/appliance_installation_script.rb.erb [hunk 3]
```diff
@@ -7,22 +7,34 @@ require "bundler/definition"
...
+<%= gem[:env_name] %>_INSTALL_LOCATION = "#{GEM_INSTALL_DIR}/gems/#{<%= gem[:env_name] %>_FULL_GEM_NAME}"
+<% end -%>
+<% end -%>
+
 VMDB_DIR                 = "/var/www/miq/vmdb"
 
+<% installed_gem = @solo_gem ? "#{@other_gems.first[:name]}.gem" : "manageiq-performance.gem" -%>
 # The following string is a base64 encoded string of the
-# manageiq-performance.gem file, and was generated by using the following rake
+# <%= installed_gem %> file, and was generated by using the following rake
```
Comment (kbrock) [inline tasks/support/templates/appliance_installation_script.rb.erb]: feels like there are so many variables here :(

**Comment by NickLaMuro** [comment]
@kbrock I actually plan on doing the `TODO` section in the PR description before shipping this, so this is definitely on in it's final form yet.  I actually have some tests in place which I can push up, but then I still want to get the other parts in place as well (the `_c_ext` tasks), so there is a bit more to do yet on this one before I ship.

**Comment by NickLaMuro** [comment]
@kbrock just pushed what I had, and I think it at least adds tests around what I had.  Still will have some more changes coming, but you should see significant differences from what you reviewed earlier.

**Comment by NickLaMuro** [comment]
@kbrock I think this is ready for review now.  Rebased out (well... technically `git branch -m`'d the old branch and cherry picked in the needed commits on top of master...) the changes from https://github.com/ManageIQ/manageiq-performance/pull/39 so this should be a little more clear to review.

I think the following are new since you have looked:

- https://github.com/ManageIQ/manageiq-performance/pull/25/commits/997ae221a6eee062fc489e3e3ed8735f996fa4d1
- https://github.com/ManageIQ/manageiq-performance/pull/25/commits/c565f7b134a3084d9feeef2db728f33164baebeb
- https://github.com/ManageIQ/manageiq-performance/pull/25/commits/fe288a958facd186773e145018d31829bd31d752
- https://github.com/ManageIQ/manageiq-performance/pull/25/commits/205d30ed9c71b8bfe3b08692d7e2c77c1bbc7492

File: tasks/support/templates/appliance_installation_script.rb.erb (hunk 3 above)
Comment (kbrock) [reply tasks/support/templates/appliance_installation_script.rb.erb]: much better

File: tasks/support/templates/appliance_installation_script.rb.erb [hunk 4]
```diff
@@ -36,34 +39,33 @@ VMDB_DIR                 = "/var/www/miq/vmdb"
...
 #     https://github.com/pypa/get-pip/blob/master/tasks/generate.py
 #
-MIQPERF_GEM_BUNDLE = <<-GEM
-<%= @gem_base64_string -%>
-GEM
+<% unless @solo_gem -%>
+<%= render_partial "gem_base64_string", :env_name => "MIQPERF", :gem_base64_string => @gem_base64_string -%>
+<% end -%>
 
 <% if @stackprof_gemspec and @stackprof_gem_base64_string -%>
```
Comment (kbrock) [inline tasks/support/templates/appliance_installation_script.rb.erb]: was expecting to see a `!@solo_gem` here.

maybe that is built into `@stackprof_gemspec`?

File: tasks/support/templates/appliance_installation_script.rb.erb [hunk 5]
```diff
@@ -36,34 +39,33 @@ VMDB_DIR                 = "/var/www/miq/vmdb"
...
-STACKPROF_GEM_BUNDLE = <<-STACKPROF_GEM
-<%= @stackprof_gem_base64_string -%>
-STACKPROF_GEM
+<%= render_partial "gem_base64_string", :env_name => "STACKPROF", :gem_base64_string => @stackprof_gem_base64_string -%>
 <% end -%>
 
+<% if @other_gems -%>
```
Comment (kbrock) [inline tasks/support/templates/appliance_installation_script.rb.erb]: one day - would be cool if all gems were treated the same.

File: tasks/support/templates/appliance_installation_script.rb.erb (hunk 4 above)
Comment (NickLaMuro) [reply tasks/support/templates/appliance_installation_script.rb.erb]: @kbrock as it is right now, `@stackprof` and `@solo_gem` are triggered by different parts of the codebase, even though they are fundamentally the same.

For example this:

```console
$ rake extra_c_gem[stackprof] generate_install_script
```

Would generate basically the same thing as:

```console
$ rake include_stackprof generate_install_script
```

That said, consider these:

```console
# installs `rbtrace`
$ rake solo_c_gem[rbtrace] generate_install_script
```

```console
# installs `stackprof` and `rbtrace`
$ rake include_stackprof solo_c_gem[rbtrace] generate_install_script
```

```console
# installs `stackprof` twice... and probably fails
$ rake include_stackprof solo_c_gem[stackprof] generate_install_script
```

Probably poorly named, but `@solo_gem` is more meant to be "create a script that doesn't include `manageiq-performance`", and not "only install this one gem".  Naming is hard...

File: tasks/support/templates/appliance_installation_script.rb.erb (hunk 5 above)
Comment (NickLaMuro) [reply tasks/support/templates/appliance_installation_script.rb.erb]: Agreed.  On the TODO list for sure, but for now, want to close this out.  :shipitkeenan:

File: tasks/support/templates/appliance_installation_script.rb.erb (hunk 4 above)
Comment (NickLaMuro) [reply tasks/support/templates/appliance_installation_script.rb.erb]: Also, as mentioned under the TODO items:

> ~~* [ ] Have the `include_stackprof` task call out to a generic `extra_c_gem` task~~
> 
> Will do this last one in another pass I think.  Would rather have tests in place before doing that.

So I do have plans to make `stackprof` quack like everything else, and just have the helper task there for convenience.

**Pull request merged**